runtime/src/accounts_db.rs

```diff
@@ -2522,44 +2523,57 @@ impl AccountsDb {
         // Working from the beginning of store_usage which are the most sparse and see when we can stop
         // shrinking while still achieving the overall goals.
         let mut shrink_slots: ShrinkCandidates = HashMap::new();
+        let mut shrink_slots_next_batch: ShrinkCandidates = HashMap::new();
         for usage in &store_usage {
+            let store = &usage.store;
             let alive_ratio = (total_alive_bytes as f64) / (total_bytes as f64);
+            debug!("alive_ratio: {:?} store_id: {:?}, store_ratio: {:?} requirment: {:?}, total_bytes: {:?} total_alive_bytes: {:?}",
+                alive_ratio, usage.store.append_vec_id(), usage.alive_ratio, shrink_ratio, total_bytes, total_alive_bytes);
             if alive_ratio > shrink_ratio {
                 // we have reached our goal, stop
                 debug!(
                     "Shrinking goal can be achieved at slot {:?}, total_alive_bytes: {:?} \
                     total_bytes: {:?}, alive_ratio: {:}, shrink_ratio: {:?}",
                     usage.slot, total_alive_bytes, total_bytes, alive_ratio, shrink_ratio
                 );
-                break;
+                if usage.alive_ratio < shrink_ratio {
```

Comment:
also it's worth noting dead slots are removed from the shrink candidate set here: https://github.com/solana-labs/solana/blob/4d91d6922149725452dbc687b398e60bca8902d8/runtime/src/accounts_db.rs#L2202

Reply:
This shrink candidate set should be bounded by the number of alive bytes satisfying this condition here: https://github.com/solana-labs/solana/blob/4d91d6922149725452dbc687b398e60bca8902d8/runtime/src/accounts_db.rs#L2531-L2534. 

I.e. the only way  for the shrink candidate set to keep growing in size is if:
1.  there are a bunch of little stores that all satisfy `usage.alive_ratio < shrink_ratio`
2.  None of those little stores are dead, so still remain in the shrink candidate set.
3.  `alive_ratio > shrink_ratio` for the entire system, even without cleaning these little stores.

If zero lamport account cleaning is working properly, all slots should eventually be marked dead within an epoch so that should bound the number of candidates to 430K. The only time this wouldn't happens is if we have a chain of zero lamport accounts that takes *multiple* cleans to purge because they are mutually dependent.

In this latter case if zero-lamport accounts are not being cleaned up, I think the shrink candidate set can in worst case be at least as large as the number of zero-lamport keys, which in degenerate cases as we saw this week, could be millions

Reply:
Thanks for the analysis Carl!